### PR TITLE
CRM-15619

### PR DESCRIPTION
**Pull request by kurund**
https://issues.civicrm.org/jira/browse/CRM-15619

### CODE REVIEW

File: tests/phpunit/api/v3/SyntaxConformanceTest.php [hunk 1]
```diff
@@ -1057,6 +1057,12 @@ public function testCreateSingleValueAlter($entityName) {
             $optionValue = $this->callAPISuccess('option_value', 'create', array('option_group_id' => $optionGroupID, 'label' => 'new option value'));
             $options['values'][] = $optionValue['id'];
           }
+          if (isset($specs['pseudoconstant']['table'])) {
+            //some psedoconstant property is based on other entity attributes
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/SyntaxConformanceTest.php]: I'm not sure about this - perhaps @totten might have an opinion - but I think we are already testing some non-option-value pseudoconstants here & this reduces our test coverage of those. In theory there is no reason for non-option-value pseudoconstatnts not to work as the createTestObject should create them - although I'm not going to argue that it necessarily does what it should do.

File: tests/phpunit/api/v3/ContributionTest.php [hunk 2]
```diff
@@ -1157,8 +1151,8 @@ public function testCreateUpdateContribution() {
 
     $this->assertEquals($contribution['values'][$contributionID]['contact_id'], $this->_individualId);
     $this->assertEquals($contribution['values'][$contributionID]['total_amount'], 110.00);
-    $this->assertEquals($contribution['values'][$contributionID]['financial_type_id'], $this->_financialTypeId);
-    $this->assertEquals($contribution['values'][$contributionID]['instrument_id'], $old_payment_instrument);
+    $this->assertEquals($contribution['values'][$contributionID]['financial_type_id'],$this->_financialTypeId );
+    $this->assertEquals($contribution['values'][$contributionID]['payment_instrument_id'], $old_payment_instrument);
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/ContributionTest.php]: Unfortunately this breaks our 'test contract' - ie. that we will continue to support tested inputs & outputs in the api. In other cases we have added a function to the api to handled things like this e.g 

/**
 * Support for schema changes made in 4.2
 * The main purpose of the API is to provide integrators a level of stability not provided by
 * the core code or schema - this means we have to provide support for api calls (where possible)
 * across schema changes.
 */
function _civicrm_api3_event_create_legacy_support_42(&$params){
  if(!empty($params['payment_processor_id'])){
    $params['payment_processor'] = CRM_Core_DAO::VALUE_SEPARATOR . $params['payment_processor_id'] . CRM_Core_DAO::VALUE_SEPARATOR;
  }
}

File: tests/phpunit/api/v3/ContributionTest.php [hunk 3]
```diff
@@ -456,21 +456,15 @@ public function testCreateContributionWithPaymentInstrument() {
   public function testGetContributionByPaymentInstrument() {
     $params = $this->_params + array('payment_instrument' => 'EFT');
     $params2 = $this->_params + array('payment_instrument' => 'Cash');
-    $this->callAPISuccess('contribution', 'create', $params);
-    $this->callAPISuccess('contribution', 'create', $params2);
-    $contribution = $this->callAPISuccess('contribution', 'get', array(
-        'sequential' => 1,
-        'contribution_payment_instrument_id' => 'Cash'
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/ContributionTest.php]: 'contribution_payment_instrument_id' needs to continue to work as part of our test contract - in this case I would think it still would as it is the unique name for the field - but we shouldn't 'tidy the test for it' such that the existing contract is no longer tested

File: tests/phpunit/api/v3/UFGroupTest.php [hunk 4]
```diff
@@ -58,7 +58,7 @@ protected function setUp() {
       'uf_id' => 42,
       'uf_name' => 'email@mail.com',
     ));
-    $this->_ufGroupId = $ufMatch['id'];
+    $this->_ufMatchId = $ufMatch['id'];
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/UFGroupTest.php]: phenonemal :-)

File: api/v3/utils.php [hunk 5]
```diff
@@ -1931,17 +1998,23 @@ function _civicrm_api3_api_match_pseudoconstant(&$params, $entity, $fieldName, $
     }
     $options = civicrm_api($entity, 'getoptions', array('version' => 3, 'field' => $fieldInfo['name'], 'context' => 'validate'));
     $options = CRM_Utils_Array::value('values', $options, array());
+
+    if (count($options) == 0 && isset($pseudoconstant['table'])) {
+        $pseudoParams = $pseudoconstant;
+        unset($pseudoParams['table']);
+        $options = CRM_Core_PseudoConstant::get(CRM_Core_DAO_AllCoreTables::getClassForTable($pseudoconstant['table']), $fieldName, $pseudoParams);
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: @colemanw is there a problem with getoptions that is being worked around here?

File: api/v3/utils.php [hunk 6]
```diff
@@ -1883,35 +1926,57 @@ function _civicrm_api3_validate_html(&$params, &$fieldName, $fieldInfo) {
...
+  
+  if ($fieldValue) {
+    if (!CRM_Utils_Rule::xssString($fieldValue)) {
+      throw new Exception('Illegal characters in input (potential scripting attack)');
     }
     if ($fieldName == 'currency') {
-      if (!CRM_Utils_Rule::currencyCode($value)) {
-        throw new Exception("Currency not a valid code: $value");
+      //When using IN operator $fieldValue is a array of currency codes
+      if (is_array($fieldValue)) {
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: just a note - a trick Coleman taught me for dealing with not knowing if you have a string or an array is 

foreach ((array) $fieldValue as $currency) {

}

it can reduce code lines - which usually improves readability

File: api/v3/ReportTemplate.php [hunk 7]
```diff
@@ -109,7 +109,7 @@ function _civicrm_api3_report_template_getrows($params) {
   }
 
   $class = civicrm_api3('option_value', 'getvalue', array(
-    'option_group_id' => 'report_template',
```
Comment (eileenmcnaughton) [inline api/v3/ReportTemplate.php]: why was this changed? was if preference or because it stopped working? The api should continue to support strings for option_group_id so hopefully it was the former

**Comment by eileenmcnaughton** [comment]
565 additions and 688 deletions = +123 karma points :-)

Looks like a lot of standardisation in this patch

File: api/v3/utils.php (hunk 5 above)
Comment (colemanw) [inline api/v3/utils.php]: I have no idea. @monishdeb could you please add code comments here so we can understand why it is necessary.

Comment (monishdeb) [inline api/v3/utils.php]: Those attribute(s) which has pseudo property as 'table' and specific keyColumn/LabelColumn value didn't fall under 'getoptions' api. Like in xml you will find that I have added pseudoconstant for contrubution_soft.pcp_id
<pseudoconstant>
+      <table>civicrm_pcp</table>
+      <keyColumn>id</keyColumn>
+      <labelColumn>title</labelColumn>
+    </pseudoconstant>
So no options (id=>title) get retrieved via getoptions.api so I handled it in here to support 'pseudoconstant' property too not ONLY 'option'

  I'll surely add a comment above this code.

Comment (eileenmcnaughton) [inline api/v3/utils.php]: Are you saying that 

civicrm_api3('contribution_soft', 'getoptions', array('field' => 'pcp_id')) doesn't return anything? Or that what it returns is the 'wrong' field for the purpose eg. returning 'name' rather than 'label'?

Comment (colemanw) [inline api/v3/utils.php]: I want to clarify that api.getoptions *does* understand and should correctly handle pseudoconstants that specify "table" and *does not* require the corresponding BAO to have a "buildOptions" method (since all BAOs inherit this method from CRM_Core_DAO it is optional but not necessary for them to override it).

So I still don't understand the problem you are trying to solve, but I suspect this is not the correct solution. If api.getoptions is messing up we should fix it rather than working around it.
